### ISSUE TITLE
Is there a better way to handle idling on multiple tabs:

### ISSUE THREAD

**Issue by sandipsubedi** [SupremeTechnopriest/react-idle-timer#147]
Basically, I need to logout the user if they are idle on all tabs. And not log them out if they are active in even 1 tab.

This is how I am doing now. But I am updating my localStorage every time the user is doing something. It feels like there should be a better way to do this. Anyone ran into this before and have any suggestions?

```js
  const timeout = Number(USER_IDLE_TIMEOUT) * 1000
  const handleOnIdle = () => {
    const lastActive = Number(localStorage.getItem("lastActive"))
    if (Date.now() > lastActive + timeout) {
      logOut()
    }
  }

  const handleOnActive = () => {
    // This is to handle user's activity on multiple tabs.
    localStorage.setItem("lastActive", `${Date.now()}`)
  }

  useIdleTimer({
    timeout: timeout,
    onIdle: handleOnIdle,
    onActive: handleOnActive
  })

```

Another way I found was to have a counter for tab and idleTab. This way last idle tab will logout user from all tabs

```js
  const timeout = Number(USER_IDLE_TIMEOUT) * 1000
  const handleOnIdle = () => {
    let idleCount:number = Number(localStorage.getItem("idleCount")) || 0
    idleCount++
    localStorage.setItem("idleCount", `${idleCount}`)
    let tabCount:number = Number(localStorage.getItem("tabCount")) || 0
    // Don't do anything if user is not logged in:
    if (idleCount == tabCount) {
      logOut()
    }
  }

  useIdleTimer({
    timeout: timeout,
    onIdle: handleOnIdle,
  })

  useEffect(()  => {
    window.onload = () => {
      let tabCount:number = Number(localStorage.getItem("tabCount")) || 0
      tabCount++
      localStorage.setItem('tabCount',`${tabCount}`)
      const channel = new BroadcastChannel('logout');
      channel.onmessage = (msg: any) =>{
        if(msg === 'logout'){
          window.location.href = "/"
        }
      } 
    }

    window.onbeforeunload = () => {
      let tabCount:number = Number(localStorage.getItem("tabCount")) || 0
      tabCount--
      localStorage.setItem('tabCount', `${tabCount}`)
    }
  })
```

**Comment by SupremeTechnopriest** [comment]
So you can actually `postMessage` in between tabs.  You could designate one tab as your orchestrator, and then have all other tabs postMessage to the orchestrator tab.  Would look something like this:

1. First tab is opened, designate as orchestrator
2. New tab comes up, postMessage to orchestrator to say "Hello I'm here". Orchestrator adds tab to a map with idle state
3. On Idle timer (active, idle) events, if you aren't the orchestrator, postMessage to orchestrator and update the state of the tab
4. Every time the orchestrator updates the state, check that there is at least one active tab. If not log out.
5. Then before the page unloads you tell the orchestrator to remove the tab from state.  

This would prevent you from having to use storage and you can just keep the state in memory.  This question seems to come up quite a bit, maybe I will see if I can implement it in the library or at least, make an example of how to do it.

**Comment by SupremeTechnopriest** [comment]
On second thought, maybe I can build this into the library using a SharedWorker.... That would be pretty elegant.  I will experiment with it over the next few days and let you know.

**Comment by sandipsubedi** [comment]
> So you can actually postMessage in between tabs. You could designate one tab as your orchestrator, and then have all other tabs postMessage to the orchestrator tab. Would look something like this:

I didn't know about this before. I will give it a shot. Thank you so much for your quick response.

**Comment by sandipsubedi** [comment]
>On second thought, maybe I can build this into the library using a SharedWorker.... That would be pretty elegant. I will experiment with it over the next few days and let you know.

It seems like SharedWorker is not supported on Internet Explorer (surprise 😆 ) and no full support on Safari. So that's probably a no go for me.

**Comment by SupremeTechnopriest** [comment]
Sounds good.  Good luck.  Im going to think about a way to add this functionality to the core for the next major release.

**Comment by sandipsubedi** [comment]
Thank you @SupremeTechnopriest . I am going to close.

**Comment by SupremeTechnopriest** [comment]
Hey @sandipsubedi,

I am pretty close to finishing a built in implementation for multiple tabs and I had a few questions.

- When all tabs are idle, would you want each tab to fire the `onIdle` event? Or just the last or currently visible tab?
- When a tab becomes active, would you want `onActive` to be fired on all tabs, just the active tab, or the last visible or currently tab? 

My thoughts are, maybe the `onIdle` event opens an "Are you still there?" model, in which case you probably want it fired on all tabs, but if you are directly logging the user out with `onIdle`, you might only want one tab to fire the event?  Maybe I make this a configurable option?

What do you think?

Also, I looked at various different methods to achieve this and I ended up going with `localStorage` as well (for now).  I think the best API to use would be the `BroadcastChannel` API, but it's not supported in IE or Safari.  Maybe one day when IE is actually purged from the wild and Safari catches up with the rest of the world we can make the switch. 

The downside of using `localStorage` is that setting storage blocks the main thread.  The `BroadcastChannel` would be non blocking.  With postMessage it's difficult to get a reference to the other tab windows and `SharedWorkers` have the same limitations as `BroadcastChannels`.

**Comment by sandipsubedi** [comment]
>When all tabs are idle, would you want each tab to fire the onIdle event? Or just the last or currently visible tab?

In my case, the last one is better. Because I want to logout the user when the `last tab that's active` is idle.

**Comment by sandipsubedi** [comment]
> When a tab becomes active, would you want onActive to be fired on all tabs, just the active tab, or the last visible or currently tab?

In my case, I care more about all tabs being inactive for let's say 30 min. And the last tab that becomes inactive for 30min will check if there is still an active tab. If there is, it won't logout the user. If all other tabs are inactive for 30min already, I want to logout the user.

If the user is active, it probably has to update some central place that it's still active, so that the current tab knows if there is still an active tab so that it doesn't logout the user. 

For Broadcasting I am using this: https://github.com/pubkey/broadcast-channel and according to the documentation, it seems like it's IE compatible.

**Comment by sandipsubedi** [comment]
> My thoughts are, maybe the onIdle event opens an "Are you still there?" model, in which case you probably want it fired on all tabs, but if you are directly logging the user out with onIdle, you might only want one tab to fire the event? Maybe I make this a configurable option?

A couple of things here. I am already using this: https://github.com/pubkey/broadcast-channel  to broadcast a message from my **logout method** that will logout user from all tabs (technically user is logged out from one tab and other tabs will be redirected to the login page/home page). 

Another piece is,
`My thoughts are, maybe the onIdle event opens an "Are you still there?" model, in which case you probably want it fired on all tabs,`, This is nice in a way to check if a user is active on another tab. But if the firing of even is there and I can even utilize that to call logout method so that I can logout the user, that's even better. At that point, I probably don't even have to use that other package.

Thank you so much again for creating this package and spending so much time adding new features 🙏

**Comment by SupremeTechnopriest** [comment]
Okay, I think I will make it configurable to cover all use cases. I hope to have this ready to test by the end of the month.

**Comment by sandipsubedi** [comment]
Btw, storing `tabCount` on localStorage wasn't very reliable while redirecting. But I ended up doing it as you suggested above, using broadcast message and making one tab the elector/leader.


```js
  const handleOnIdle = () => {
    if (window.location.pathname !== "/") {
      let idleCount: number = Number(localStorage.getItem("idleCount")) || 0
      localStorage.setItem("tabCount", "0")
      const channel = new BroadcastChannel("tabCountChannel")
      // Setting this tab as the leader.
      const elector = createLeaderElection(channel)
      channel.postMessage("tabCount")
      idleCount++
      localStorage.setItem("idleCount", `${idleCount}`)
      elector.awaitLeadership().then(() => {
        const tabCount = Number(localStorage.getItem("tabCount")) || 0
        const idleCount: number = Number(localStorage.getItem("idleCount")) || 0
        // Don't do anything if user is not logged in:
        // Logout user from the last inactive tab.
        if (l idleCount >= tabCount) {
          logOut()
        }
        elector.die()
      })
    }
  }


  useIdleTimer({
    timeout,
    onIdle: handleOnIdle
  })


  useEffect(() => {
    window.onload = () => {
      const logoutChannel = new BroadcastChannel("logoutChannel")
      const tabCountChannel = new BroadcastChannel("tabCountChannel")
      logoutChannel.onmessage = (msg: string) => {
        if (msg === "logout") {
          localStorage.clear()
          const logOutUrl = MY_URI
          window.location.href = logOutUrl
        }
      }
      tabCountChannel.onmessage = (msg: string) => {
        if (msg === "tabCount") {
          let tabCount: number = Number(localStorage.getItem("tabCount")) || 0
          tabCount++
          localStorage.setItem("tabCount", `${tabCount}`)
        }
      }
    }
  })

```

**Comment by SupremeTechnopriest** [comment]
Cool cool!  Glad it helped.  Work is a bit busy at the moment so it's been a bit tough to find the time to work on this, but I have it mostly reliable on a new branch. After a few more iterations, I'll probably push out a `next` ask you to test it.

**Comment by sandipsubedi** [comment]
Awesome! I will be more than happy to test it.

**Comment by Niks218** [comment]
I found out this topic and created my own snippet. Maybe it will help you and everyone who will try to implement something similar
```
const { reset, pause } = useIdleTimer({
    timeout: LOGIN_SESSION_DURATION,
    throttle: 1000 * 60,
    onAction: handleAction,
    startOnMount: false,
  });

  function setupBroadcastChannel(): [elector: LeaderElector, channel: BroadcastChannel<'reset'>] {
    const handler: OnMessageHandler<'reset'> = () => reset();

    const channel = new BroadcastChannel<'reset'>('timer');

    const elector = createLeaderElection(channel, { fallbackInterval: 2000, responseTime: 100 });
    elector.awaitLeadership().then(() => {
      reset();
      channel.addEventListener('message', handler);
    });
    elector.onduplicate = () => elector.die();

    channel.postMessage('reset');

    return [elector, channel];
  }

  useEffect(() => {
    if (isLoggedIn) {
      const [elector, channel] = setupBroadcastChannel();

      return () => {
        elector
          .die()
          .then(() => channel.postMessage('reset'))
          .then(() => channel.close());
      };
    } else {
      pause();
    }
  }, [isLoggedIn]);
```

**Comment by SupremeTechnopriest** [comment]
Nice looks good @Niks218!  For my implementation I wont be using BroadcastChannel just yet, as they aren't [widely supported](https://caniuse.com/?search=BroadcastChannel), but the interface of the API can easily be swapped out for BroadcastChannel when support is more uniform in the wild. Anyone who doesn't need to support IE, Safari or iOS I strongly recommend using BroadcastChannel as its non-blocking.

**Comment by SupremeTechnopriest** [comment]
I imagine you are using this? https://github.com/pubkey/broadcast-channel

Not bad, I will look into it.  It has nice fallbacks for older browsers.

**Comment by Niks218** [comment]
@SupremeTechnopriest Yes, I used this library because @sandipsubedi mentioned it. I don't care much about IE, but Safari is something that should be supported. So yeah, didn't want to create something unmaintainable.
Also, I tried to implement your idea with orchestrator tab and this **broadcast-channel** library came to rescue with its mechanism for leader selection. Made work much easier.
The only drawback I noticed, your timer can't be fully controllable, so every tab anyway has running timer. Maybe, it would be good to has some exposed method like `start` or something IDK.
Great library BTW, thank you!

**Comment by SupremeTechnopriest** [comment]
You have `pause` and `resume` to control the timer 🙂. I will get a robust cross tab solution in place so you can just set it and forget it.  I think I'll go with this broadcast channel lib.  It looks like it will do nicely. Currently messing around with some ideas.  I will post back here with configuration examples and get your feedback! Thanks for the snippet!

**Comment by Niks218** [comment]
That's true, but that's less efficient then no to start a timer at all IMO. Anyway it's not a big problem for me.

Great! Definitely will try and switch to your solution.

**Comment by SupremeTechnopriest** [comment]
Ah! You can set `startOnMount` to `false` and the timer wont be started until you call `reset()`.  Would that do what you need?

**Comment by Niks218** [comment]
Hmm, I tried it, but timer was started anyway after some action. Most likely, I messed up somewhere.

**Comment by SupremeTechnopriest** [comment]
You are right, setting that to false would make it start on first action.  I will add something for you to manually start the timer.

**Comment by Niks218** [comment]
It would be very cool, thank you a lot!

**Comment by dreaken** [comment]
Any progress in this area? :)

**Comment by SupremeTechnopriest** [comment]
Hey all, sorry for the delay here.  Work has been really busy with a new major roll out.  Im going to be wrapping this up this weekend along with some other new additions.  I will keep you posted!

**Comment by SupremeTechnopriest** [comment]
Hey @sandipsubedi, @Niks218 and @dreaken!

So I have `4.6.0-rc.0` published to npm under the `next` tag.  You can install it with `npm i react-idle-timer@next`.

Right now I have only implemented cross tab support for the `useIdleTimer` hook, but once we all agree that is stable I will port it over to the component as well.  Here's how to use it:

```javascript
useIdleTimer({ crossTab: true })
```

It currently defaults to off. You can also supply an object config to the `crossTab` property.  I have updated the typescript definitions, but here's a breakdown:

```typescript

export type ManagerType = 'broadcastMessage' | 'localStorage' | 'simulate'

interface CrossTabInterface {
    /**
     * Manager type. Selected automatically if left undefined. 
     */
    type?: ManagerType,

    /**
     * Broadcast Channel name. Default: idle-timer.
     */
    channelName?: string,

    /**
     * How often renegotiation for leader will occur. Default: 2000.
     */
    fallbackInterval?: number,
    
    /**
     * How long tab instances will have to respond. Default: 100.
     */
    responseTime?: number
    
    /**
     * LocalStorage item time to live. Default: 60000.
     */
    removeTimeout?: number
  }
```

I ended up pulling in the `broadcast-channel` code locally and I stripped out all the functionality we don't need to keep things as light as possible. This will elect a leader and when all tabs reach a synchronized state, the leader will emit the `onActive` and `onIdle` events.

I also added a `startManually` option to the config and an alias to `reset()` called `start()`.  This will allow you to start the timer when ever you want. You can call either `reset` or `start`. They do the same thing, I just made the alias so things are more semantic.  You can stop the timer in its tracks with `pause` as usual. 

```javascript
const { start } = useIdleTimer({ startManually: true })
start()
```

If you all can run this through the paces that would be excellent.  I'm open to any feedback and we will use this issue for any bug reports. Once we all feel it's in a good place, I'll back port to `<IdleTimer />`, write some docs and deploy this.  I tried writing some tests for it today, but the async nature of all this makes it pretty difficult to test reliably. Im probably going to just unit test the individual components and write a few small tests for integration, but will work on that once we get it tested by hand in a few apps. 

Let me know how it goes!

**Comment by dreaken** [comment]
As now im not able to write code yet, i have question. OnActive will be emited to every tabs automatically, or manual? I want to display login layer on every tabs, and hide it, if user authorize on one od them.

**Comment by SupremeTechnopriest** [comment]
@dreaken Ah in its current form it will not do that, but I can easily add that functionality.  I'll put it behind a config option. I will push rc1 today for you to test.

**Comment by SupremeTechnopriest** [comment]
@dreaken I just published `4.6.0-rc.1` to the `@next` tag.  You can now supply an `emitOnAllTabs` option to have `onIdle` and `onActive` emitted on all tabs. Typedefs updated as well.  Here is an example:

```javascript
useIdleTimer({
  crossTab: {
    emitOnAllTabs: true
  }
}
```

This will fire the events on all tabs when all tabs become idle or when one tab becomes active from all tabs being idle.

Let me know how it goes!

**Comment by SupremeTechnopriest** [comment]
Has anyone had a chance to test this new code in their application yet?  Im hoping to release it by the end of the week.  It all looks good to me, but it would be nice to verify that it works as expected in a full application and not just my little sandboxes.

**Comment by sandipsubedi** [comment]
Hi, @SupremeTechnopriest  I will try this today and let you know.

**Comment by Niks218** [comment]
@SupremeTechnopriest I tried it and it work great. But I noticed, that when you do some actions on non-leader tab, then it resets shared timer, but doesn't call `onAction` callback at others tabs or at least at leader tab. Is this done on purpose?

**Comment by SupremeTechnopriest** [comment]
@Niks218 Did you try setting `emitOnAllTabs`?

**Comment by Niks218** [comment]
Yes, I'm using it. How I said that's good for me. Works like a charm

**Comment by sandipsubedi** [comment]
@SupremeTechnopriest It's working great for me too! Good work here!

**Comment by SupremeTechnopriest** [comment]
Awesome!  Im going to work on getting this ready for release.  Outlining the steps below:

- [ ] Unit tests for TabManager
- [ ] Port functionality to IdleTimer Component
- [ ] Write documentation
- [ ] Publish Release

**Comment by dreaken** [comment]
@SupremeTechnopriest  I have created a PoC and it looks good, thanks!

**Comment by SupremeTechnopriest** [comment]
Painfully made it to 100% test coverage.  On to docs and examples.

**Comment by SupremeTechnopriest** [comment]
Hey everyone.  `4.6.0` is deployed to npm.  Check out the README for full docs and the CHANGELOG or release for changes.  Thank you all for testing this feature with me!

**Comment by dreaken** [comment]
@SupremeTechnopriest I have one issue, how i can switch from idle, to active? I have useIdleTimer initialized in this way: 
![image](https://user-images.githubusercontent.com/6353443/114287734-95c6a800-9a69-11eb-9da6-6c497f0c241f.png)
and i tried run reset/resume, but onActive was not executed. If i will attach my logic to onAction, then after reset it will be no synced with other tabs.

**Comment by SupremeTechnopriest** [comment]
Do you have:

```javascript
{
  crossTab: {
    emitOnAllTabs: true
  }
}
```

Right now, calling reset/resume only applies to the tab you call it on.  Maybe we should sync that across tabs as well?

**Comment by dreaken** [comment]
Yes 
![image](https://user-images.githubusercontent.com/6353443/114288101-d5db5a00-9a6c-11eb-9595-8c0dcb4c9d9d.png)
but onActive is not even fired on local tab.

if we are talking about synchronize reset between tabs, i think yes. If we manipulate IdleTimer, then we want to have effect in every tab

**Comment by SupremeTechnopriest** [comment]
Can you test on the 4.6.0 latest tag?  I think i know whats up.

**Comment by SupremeTechnopriest** [comment]
Yeah reset and reset don't call onActive. Only user input can call onActive.  Since you have `stopOnIdle` set onActive will never be called.  You basically have it in manual mode.  You can just do your onActive stuff when you call resume or reset.  Im not sure that firing `onActive` on reset and resume would be good for other implementations.  It would end up always firing onActive when the component mounts.  `onActive` is designed to be called when a user becomes active from an idle state.

**Comment by dreaken** [comment]
How can i run onActive manually? Cuz i need to use stopOnIdle, becouse i need to set status on active, only if authorization will pass

**Comment by SupremeTechnopriest** [comment]
I would just make it a function you call:

```javascript
const onActive = () => {
  // Your active logic
}


// Later

onActive()
reset()
```

**Comment by dreaken** [comment]
but onActive have to be fired on every tabs, becouse it hides authorization overlay

**Comment by SupremeTechnopriest** [comment]
Mmm I see.  Let me think on that.  I will have to add something to allow you to do this.

**Comment by dreaken** [comment]
It will be great! I count on you!

**Comment by SupremeTechnopriest** [comment]
I have a few ideas I will try out this weekend.  Ill get back to you here.

**Comment by dreaken** [comment]
Do you use "BroadcastChannel"? Because i get often error like this
![image](https://user-images.githubusercontent.com/6353443/114313978-15e81e80-9af9-11eb-8f3d-743bc594f083.png)

**Comment by SupremeTechnopriest** [comment]
Broadcast channel has a bunch of extra junk we don't need here.  I implemented only what we need from there.  I think i know what this is, I will fix it today along with the send onActive feature you requested.

**Comment by SupremeTechnopriest** [comment]
@dreaken just to be sure, you are on the latest tag right? I didn't push the final version to next tag.

**Comment by dreaken** [comment]
I'm using "react-idle-timer": "^4.6.0",

**Comment by SupremeTechnopriest** [comment]
Cool, leave this with me for a bit.  Ill get back to you shortly.

**Comment by SupremeTechnopriest** [comment]
@dreaken Okay, so what I did is separate out `start` and `reset`.  'start' starts the timer without firing any events and `reset` will fire `onActive` if the timer is in an idle state (no `crossTab`) or all tabs are idle (`crossTab`).  I couldn't reproduce the could not send message error.  If you can create a codesandbox that recreates that I can look into it.  All the tests pass.

You can pull the `@next` tag again and test it out.  If it works ok for you ill roll out `4.6.1`.

**Comment by dreaken** [comment]
In my reallife app it doesnt fire onActice, but on fresh, simple app, yes, but it doesnt fire reset.

For example, main tam, when i run reset:
![image](https://user-images.githubusercontent.com/6353443/114439984-e6551700-9bc9-11eb-8e7e-d8b3169fa5e4.png)

second tab
![image](https://user-images.githubusercontent.com/6353443/114440056-fbca4100-9bc9-11eb-9a2c-356b5af933a4.png)

We can see, that onActive was fired on second tab, but events was not registered after switching from idle to active in child tab

**Comment by dreaken** [comment]
I have use simple code in fresh project.
To initialize: ```
  const { getRemainingTime, getLastActiveTime, reset, start, resume } = useIdleTimer({
    timeout: 1000 * 60,
    onIdle: event => {
      console.log('user is idle', event);
      console.log('last active', getLastActiveTime());
    },
    onActive: event => {
      console.log('user is active2', event);
      console.log('time remaining', getRemainingTime());
    },
    onAction: event => {
      console.log('user is active', event);
      console.log('time remaining', getRemainingTime());
    },
    startManually: true,
    stopOnIdle: true,
    debounce: 500,
    crossTab: {
      emitOnAllTabs: true
    }
  });

  useEffect(() => start(), [])
```
![image](https://user-images.githubusercontent.com/6353443/114440389-595e8d80-9bca-11eb-92f1-df93c573b724.png)

and i just attach reset() to this text: "Edit <code>src/App.js</code> and save to reload."

**Comment by dreaken** [comment]
I also get the error in this sample project 
![image](https://user-images.githubusercontent.com/6353443/114440789-d558d580-9bca-11eb-927d-23e030070827.png)

In this case, i had 2 tabs, i did some changes(for example added second dot in "and save to reload." to get hot reload, and unactive tab crashed(not at the first time, but i added and removed this dot few times, and when i checked inactive tab i got this error)

**Comment by SupremeTechnopriest** [comment]
Okay Its probably because you aren't calling the reset function on the leader tab.  I'll push a patch where it wont matter which tab its called on. I will try to reproduce the channel closure bug as well.

**Comment by SupremeTechnopriest** [comment]
@dreaken Just pushed `4.6.1-rc.2` to the `@next` tag.  Give that a go, it should do what you are looking for now.  If `emitOnAllTabs` is enabled, all the methods will be replayed to all tabs (`start`, `reset`, `pause`, `resume`).  As far as the closed channel error. It popped up once for me, but Im pretty sure this is related to live reload.  Im going to have to get a reliable reproduction going in order to resolve it. This shouldn't happen outside of live reload.  Even with live reload on, I could only get it to happen once. Its definitely a race condition.

**Comment by dreaken** [comment]
Thanks, i will try tomorow. If we are talking about this error, in my realapp i get many times, even on reload page, so it's hard to develop if i have react-idle-timer initialized. There is no way to catch this error and try reopen channel?

**Comment by SupremeTechnopriest** [comment]
There is definitely a way to fix it (and we will), its just a matter of being able to reproduce it reliably so I can see whats going on. My guess is that the channel is being closed before the unload event is being called.  I will get into it tomorrow.

**Comment by SercanSercan** [comment]
Thank you for good work, @SupremeTechnopriest. I have a question: I don't understand the role of isLeader function. It does not change its value when the user becomes active on another tab. I read the documentation, but it is still not clear to me. Can you explain a bit more?

Thanks!

**Comment by SupremeTechnopriest** [comment]
@SercanSercan Sure!  When you have cross tab enabled, one of the tabs is elected to be the leader.  That means it is the one responsible for tracking the idle state of all other tabs and emitting events.  If the leader tab is closed, a new tab will be elected as the leader. `isLeader()` lets you know if the tab you called the function on is the lead orchestrator.  I don't really see an immediate use for the function within apps, but I use it internally and figured why not expose it in case someone needs it. 

In short, a user becoming active or idle wont effect the leader.  Only closing tabs and adding tabs will.

**Comment by SupremeTechnopriest** [comment]
@dreaken How are we looking (apart from the hot reload issue)?

**Comment by dreaken** [comment]
Sorry, i was not able to check it today, tomorrow will do for sure

**Comment by SupremeTechnopriest** [comment]
No worries.  Let me know.  I think its ready to release and there a few other fixes I would like to get out asap.

**Comment by dreaken** [comment]
Im still getting this channel error, but synchronization works fine now. thank you

**Comment by SupremeTechnopriest** [comment]
Great.  Let me see if i can fix the channel error on hot reload.

**Comment by SupremeTechnopriest** [comment]
@dreaken I published `4.6.1-rc.5` to the `@next` tag.  Instead of throwing an error when the channel is closed, I log to the console.  This will make it so the app does not crash and provide me with a little better debugging info.  Can you test it and paste the console.error here?  It will tell me which message is failing to send in time and I should be able to fix it with that info.

**Comment by dreaken** [comment]
I still get this error
![image](https://user-images.githubusercontent.com/6353443/114773779-c19aa400-9d6f-11eb-9730-9d57ce9d32eb.png)
![image](https://user-images.githubusercontent.com/6353443/114773827-d119ed00-9d6f-11eb-906f-3934cf54e0ef.png)

**Comment by SupremeTechnopriest** [comment]
Hmm something is messed up one sec.

**Comment by SupremeTechnopriest** [comment]
Is your project something you can share with me?  I cant reproduce this in my sandboxes.

**Comment by SupremeTechnopriest** [comment]
@dreaken pushed another fix attempt. `4.6.1-rc.6` can you test?

**Comment by dreaken** [comment]
i catched this error on rc5 
![image](https://user-images.githubusercontent.com/6353443/114778471-2d334000-9d75-11eb-88c3-7354cbc2d815.png)
But there is nothing interested. I think, i get this error less often than earlier. 

In my project to reproduce it, i have to open app two tabs, wait for authorization layer(cancellation request is running and one item from localstorage is cleared) then im refreshing whole page. after reload i get this error. 
Maybe it will tell you something. I will try in meantime to not clear localstorage(i think issue is someting there)

**Comment by SupremeTechnopriest** [comment]
Yeah try rc6 and see if that mitigates it. If you are using automatic method selection and you are using a modern browser, it should be using a BroadcastChannel, so localStorage wouldnt really effect anything.
 
It would be helpful if I knew which message it was trying to send when it failed.

**Comment by dreaken** [comment]
i have catch something for you
![image](https://user-images.githubusercontent.com/6353443/114785187-95395480-9d7c-11eb-9620-4d52f494e7e3.png)

**Comment by dreaken** [comment]
maybe its correlated with error above? 
![image](https://user-images.githubusercontent.com/6353443/114785490-142e8d00-9d7d-11eb-8729-851434d87bf8.png)

(but without react-idle-timer its not crashing)

**Comment by SupremeTechnopriest** [comment]
I dont think the errors are related, but that's definitely helpful.  Let me see what I can find.

**Comment by dreaken** [comment]
I think, it might be correlated, becouse if errors above are fixed, then it works

**Comment by dreaken** [comment]
which channel are these messages sent through?

**Comment by SupremeTechnopriest** [comment]
In this case it would be a BroadcastChannel called `idle-timer`.

**Comment by dreaken** [comment]
ah, yea. I thought it might be something like sockets, becouse there was one error about it

**Comment by SupremeTechnopriest** [comment]
Does the component get unmounted when it hits that error?

**Comment by dreaken** [comment]
Component with idle-timer is almost on the top of structure. i checked with simple 
![image](https://user-images.githubusercontent.com/6353443/114787581-59a08980-9d80-11eb-81c6-56109038bce8.png)
and there was no "unmounted" in the console

**Comment by dreaken** [comment]
After this error, there are still "internal" messages 
![image](https://user-images.githubusercontent.com/6353443/114788091-2a3e4c80-9d81-11eb-9901-3c3d146bef36.png)

**Comment by dreaken** [comment]
so it was able to reconnect i think. So only issue is, so the only problem I guess is that nothing catches this error

**Comment by SupremeTechnopriest** [comment]
Yeah this is the leader elector trying to elect a new leader tab.  We need to figure out why the channel is being closed.

**Comment by SupremeTechnopriest** [comment]
I can def catch the error, but the error shouldn't be occurring in the first place.

**Comment by dreaken** [comment]
True, i will try fnd out, after fixing which above error it will stop crashing

**Comment by SupremeTechnopriest** [comment]
`4.6.1`  is released to NPM.  You can switch back to the `latest` tag now.

**Comment by dreaken** [comment]
Do you catched this error in new version? Becouse even on stage serwer i get crashes becouse of it, time to time when i switch page. Additionaly, i need method, to run onActive mannualy, becouse if i get authorization layer on one page, and i open new tab with classic login page and i log in, then i am not able to hide authorization layer for previous tabs.

**Comment by dreaken** [comment]
hmm. maybe i can just use reset instead of start? I have to check it

**Comment by SupremeTechnopriest** [comment]
Yeah you can use reset for that. I didn't catch that error in this release because I thought you said it was related to another error. If that's not the case we can continue the hunt.
